### PR TITLE
Add Java Solution to Balancing(usaco-624)

### PR DESCRIPTION
**Pull request by timothygao8710**

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
this is a bit sol? thought the usaco guide sol was supposed to use binsearch but ok

**Comment by timothygao8710** [comment]
I would use binary search but I have no idea how Benq's solution works lol. Perhaps @bqi343 could provide some insight? What does one binarysearch on and how can one validate that a value works?

**Comment by bqi343** [comment]
binary search on the answer

**Comment by bqi343** [comment]
if you fix a y-coordinate and an answer, can you find the minimum and maximum x-coordinates that work?

**Comment by timothygao8710** [comment]
Hm... The editorial solution iterates through all ys and binary searches on x. How does knowing a fixed answer help? Won't you have to do the same procedure of iterating through ys and binary searching on x regardless?

**Comment by bqi343** [comment]
Consider the following simpler problem for a fixed `ans` (and assuming all coordinates are `O(N)`):

For every x-coordinate `x`, find the maximum `y` such that the number of cows `(a,b)` such that `a <= x` and `b <= y` is at most `ans`.

You should be able to do this in `O(N)`.

**Comment by timothygao8710** [comment]
Oh wait that's actually humongous brain. Thanks for the explanation.

File: solutions/usaco-624.mdx [hunk 1]
```diff
@@ -266,3 +268,102 @@ int main() {
...
+
+public class balancing {
+    static final int maxY = 1_000_001;
+    public static void main(String[] args) throws IOException {
+        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+        PrintWriter pw = new PrintWriter(System.out);
+     
+        int N = Integer.parseInt(br.readLine());
+        int[][] coords = new int[N][2];
+        for(int i = 0; i<N; i++){
```
Comment (maggie-j-liu) [inline solutions/usaco-624.mdx:289]:
```suggestion
        for(int i = 0; i < N; i++){
```
add spacing around operators everywhere else, too

File: solutions/usaco-624.mdx [hunk 2]
```diff
@@ -266,3 +268,102 @@ int main() {
 	}));
 }
 ```
+</CPPSection>
+<JavaSection>
+```java
```
Comment (maggie-j-liu) [inline solutions/usaco-624.mdx:273]: use tabs instead of spaces to indent

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/usaco-624.mdx [hunk 3]
```diff
@@ -271,98 +271,84 @@ int main() {
...
+	}
+
+	static class BIT {
+
+		public int[] bit;
+
+		public BIT(int N) {
+			bit = new int[N + 1];
+		}
+
```
Comment (SansPapyrus683) [inline solutions/usaco-624.mdx:333]: wait nvm lol
327 and 329 are unnecessary lines

File: solutions/usaco-624.mdx [hunk 4]
```diff
@@ -266,3 +268,88 @@ int main() {
 	}));
 }
 ```
+</CPPSection>
+<JavaSection>
+```java
+public class balancing {
```
Comment (SansPapyrus683) [inline solutions/usaco-624.mdx:274]:
```suggestion
public class Balancing
```

File: solutions/usaco-624.mdx [hunk 5]
```diff
@@ -266,3 +268,88 @@ int main() {
...
+	static final int maxY = 1_000_001;
```
Comment (SansPapyrus683) [inline solutions/usaco-624.mdx:275]: change the name to `MAX_Y`, all caps is the convention for constants

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Review by SansPapyrus683 — APPROVED** [review]
kinda gave up lol

**Review by dongliu0426 — APPROVED** [review]

File: solutions/usaco-624.mdx [hunk 6]
```diff
@@ -11,7 +11,9 @@ The
 [official solution](http://www.usaco.org/current/data/sol_balancing_platinum_feb16.html)
 uses a segment tree, which is introduced in Gold. However, we can just binary
 search on the answer.
+<LanguageSection>
```
Comment (maggie-j-liu) [inline solutions/usaco-624.mdx:14]: move time complexity under the explanation, and add ## Explanation and ## Implementation headers

File: solutions/usaco-624.mdx [hunk 7]
```diff
@@ -266,3 +272,86 @@ int main() {
...
+
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+		PrintWriter pw = new PrintWriter(System.out);
+
+		int N = Integer.parseInt(br.readLine());
+		int[][] coords = new int[N][2];
+		for (int i = 0; i < N; i++) {
+			StringTokenizer st = new StringTokenizer(br.readLine());
+			coords[i] = new int[] { Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken()) };
```
[the remark this answers is not in the document]
Comment (timothygao8710) [reply solutions/usaco-624.mdx:289]: changed

**Review by maggie-j-liu — APPROVED** [review]

**Pull request merged**